### PR TITLE
UnitOfWork: Support serialized and mixed transactional app

### PR DESCRIPTION
**Pull request by matthiasblaesing**
###### Problem/Solution/Result
See #3192

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
Reporter acknowledged fix works, unittests run clean,  as the author I don't see what I can do.

**Comment by joschi** [comment]
@dropwizard/committers Anybody available to review the changes?

File: dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java [hunk 1]
```diff
@@ -89,14 +89,14 @@ public void beforeStart(@Nullable UnitOfWork unitOfWork) {
...
                 ManagedSessionContext.bind(session);
-                beginTransaction(unitOfWork, session);
             } catch (Throwable th) {
                 session.close();
                 session = null;
                 ManagedSessionContext.unbind(sessionFactory);
                 throw th;
             }
         }
+        beginTransaction(unitOfWork, session);
```
Comment (amr) [inline dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java:99]: The `beginTransaction` here is no longer guarded by `try {} catch {}`, I think what you might want do is catch exceptions, and if you own the session, close/clean/unbind. e.g.:

```
try {
  beginTransaction(unitOfWork, session);
} catch (Throwable th) {
  if (sessionCreated) {
    // Same logic as the previous catch block, e.g. close/unbind/etc
  }
  throw th;
}

**Comment by amr** [comment]
Just adding my 2 cents here: I actually think this is a very simple but important change. I've been maintaining a custom fork of the `@UnitOfWork` functionality in my DW projects to workaround that issue. It's on production for 2 years now. At the time the `@UnitOfWork` aspect in DW didn't also have the `sessionCreated`-related logic (and I was amused to find that the current DW implementation almost 100% matches my modifications, including the one in this PR, which means we are all probably headed in the right direction here).

Don't get me wrong please, I'd have sent a PR long time ago but I wasn't sure at the time about my implementation and how well it's gonna work for us, and I wasn't as comfortable/confident with DW as I'm now.

But reflecting back on that change, it has been running great for us in production. And I would say it is pretty much basic functionality that everyone on the team depends on. IMHO this needs to be supported.

File: dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java (hunk 1 above)
Comment (matthiasblaesing) [reply dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java:99]: Valid point, but the documentation of the `UnitOfWorkAspect` class shows the call sequence, that requires the calls to either hit `UnitOfWorkAspect#afterEnd` or `UnitOfWorkAspect#onError`. There are two call sites: 

- `io.dropwizard.hibernate.UnitOfWorkAwareProxyFactory.create(Class<T>, Class<?>[], Object[])`
- `io.dropwizard.hibernate.UnitOfWorkApplicationListener.UnitOfWorkEventListener.onEvent(RequestEvent)`

The first is pretty straight forward and will call into the `onError` method in a catch block, the latter hooks into the jersey events and exceptions in the jersey process should result in the `ON_EXCEPTION` event being generated and invoking `onError`

Comment (amr) [reply dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java:99]: Awesome, that makes sense.

**Review by joschi — APPROVED** [review]
LGTM.

@matthiasblaesing Thanks for your contribution!

**Comment by matthiasblaesing** [comment]
@joschi  Thank you for the review!

**Comment by aceArt-GmbH** [comment]
Do all projects broken by `java.lang.IllegalStateException: Existing session transaction state (false) does not match requested (true)` after #2913 work again after this PR?

**Comment by joschi** [comment]
@aceArt-GmbH There's only one way to know: Try it. 😉

**Comment by aceArt-GmbH** [comment]
> @aceArt-GmbH There's only one way to know: Try it. wink

Sadly it does not look like it.
I don't get the exact exception anymore, but other errors taking down my application

```log
ERROR [2023-07-18 13:20:38,668] org.hibernate.engine.jdbc.spi.SqlExceptionHelper: Lock wait timeout exceeded; try restarting transaction
ERROR [2023-07-18 13:20:38,668] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: f8c733aef5a4b6f0
! com.mysql.cj.jdbc.exceptions.MySQLTransactionRollbackException: Lock wait timeout exceeded; try restarting transaction
! at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:124)
! at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
! at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:916)
! at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1061)
! at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1009)
! at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1320)
! at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:994)
! at jdk.internal.reflect.GeneratedMethodAccessor51.invoke(Unknown Source)
! at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
! at java.base/java.lang.reflect.Method.invoke(Method.java:568)
! at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:118)
! at jdk.proxy2/jdk.proxy2.$Proxy112.executeUpdate(Unknown Source)
! at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
! ... 92 common frames omitted
! Causing: org.hibernate.PessimisticLockException: could not execute statement
! at org.hibernate.dialect.MySQLDialect$3.convert(MySQLDialect.java:547)
! at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
! at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
! at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
! at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
! at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
! at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3571)
! at org.hibernate.persister.entity.AbstractEntityPersister.updateOrInsert(AbstractEntityPersister.java:3438)
! at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3852)
! at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:201)
! at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
! at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
! at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
! at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
! at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
! at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
! at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
! at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1407)
! ... 79 common frames omitted
! Causing: javax.persistence.PessimisticLockException: could not execute statement
! at org.hibernate.internal.ExceptionConverterImpl.wrapLockException(ExceptionConverterImpl.java:273)
! at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:108)
! at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
! at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
! at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1411)
! at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:489)
! at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3290)
! at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2425)
! at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:449)
! at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:183)
! at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:40)
! at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:281)
! at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101)
! at io.dropwizard.hibernate.UnitOfWorkAspect.commitTransaction(UnitOfWorkAspect.java:215)
! at io.dropwizard.hibernate.UnitOfWorkAspect.afterEnd(UnitOfWorkAspect.java:108)
! at io.dropwizard.hibernate.UnitOfWorkAwareProxyFactory.lambda$create$0(UnitOfWorkAwareProxyFactory.java:84)
```

pretty sure those problems did not appear on dropwizard 1.x

**Comment by matthiasblaesing** [comment]
@aceArt-GmbH this looks like a dead-lock in the DB. In the past their might have been a different timing and given that the transaction was prematurely committed and essentially switched following DB actions to autocommit mode might have masked your problem.
If you can create a minimal reproducer, I'm willing to have a look.

**Pull request merged**